lib/Runtime/Library/JavascriptBigInt.cpp

```diff
@@ -113,26 +156,182 @@ namespace Js
+            result->m_digits[i] = JavascriptBigInt::AddDigit(result->m_digits[i], carry, &tempCarry);
+            carry = tempCarry;
+        }
+        if (carry > 0) //increase length
+        {
+            if (result->m_length >= result->m_maxLength && !result->Resize(result->m_length + 1))
+            {
+                AssertOrFailFastMsg(false, "AbsoluteIncrement overflow");
+            }
+            result->m_digits[result->m_length++] = carry;
+        }
+        return result;
+    }
+
+    JavascriptBigInt * JavascriptBigInt::AbsoluteDecrement(JavascriptBigInt * pbi)
+    {
+        JavascriptBigInt* result = pbi;
+        Assert(!JavascriptBigInt::IsZero(result));
+        digit_t borrow = 1;
+        for (digit_t i = 0; i < result->m_length && borrow > 0; i++)
+        {
+            digit_t tempBorrow = 0;
+            result->m_digits[i] = JavascriptBigInt::SubDigit(result->m_digits[i], borrow, &tempBorrow);
+            borrow = tempBorrow;
+        }
+        Assert(borrow == 0);
+        // remove trailing zero
+        if (result->m_digits[result->m_length-1] == 0)
+        {
+            result->m_length--;
+        }
+        return result;
+    }
+
+    JavascriptBigInt * JavascriptBigInt::Increment(JavascriptBigInt * pbi)
+    {
+        if (pbi->m_isNegative)
+        {
+            // return 0n for -1n
+            if (pbi->m_length == 1 && pbi->m_digits[0] == 1)
+            {
+                JavascriptBigInt* result = pbi;
+                result->m_digits[0] = 0;
+                result->m_isNegative = false;
+                return result;
+            }
+            return JavascriptBigInt::AbsoluteDecrement(pbi);
+        }
+        return JavascriptBigInt::AbsoluteIncrement(pbi);
+    }
+
+    JavascriptBigInt * JavascriptBigInt::Decrement(JavascriptBigInt * pbi)
+    {
+        if (pbi->m_isNegative)
+        {
+            return JavascriptBigInt::AbsoluteIncrement(pbi);
+        }
+        if (JavascriptBigInt::IsZero(pbi)) // return -1n for 0n
+        {
+            JavascriptBigInt* result = pbi;
+            result->m_digits[0] = 1;
+            result->m_isNegative = true;
+            return result;
+        }
+        return JavascriptBigInt::AbsoluteDecrement(pbi);
+    }
+
+    Var JavascriptBigInt::Increment(Var aRight)
+    {
+        AssertMsg(VarIs<JavascriptBigInt>(aRight), "BigInt Increment");
+
+        JavascriptBigInt* rightBigInt = VarTo<JavascriptBigInt>(aRight);
+        return JavascriptBigInt::Increment(rightBigInt);
+    }
+
+    Var JavascriptBigInt::Decrement(Var aRight)
+    {
+        AssertMsg(VarIs<JavascriptBigInt>(aRight), "BigInt Increment");
+
+        JavascriptBigInt* rightBigInt = VarTo<JavascriptBigInt>(aRight);
+        return JavascriptBigInt::Decrement(rightBigInt);
+    }
+
+    // return low(a*b) and out high
+    digit_t JavascriptBigInt::MulDigit(digit_t a, digit_t b, digit_t* resultHigh)
+    {
+        // Multiply is performed in half chuck favor.
```

Comment:
It's good to have a portable implementation of this, but we might find that perf is improved by using platform-specific implementations. For example, on x64 when compiling with MSVC, we could use [_mul128](https://docs.microsoft.com/en-us/cpp/intrinsics/mul128?view=vs-2017).

Reply:
it fail on Linux though so I revert back to first implementation. I think we can improve it later if needed